src/java.base/share/classes/jdk/internal/foreign/abi/s390x/sysv/SysVS390xVaList.java

```diff
@@ -33,123 +33,563 @@
+	private static final long STRUCT_ARG_SIZE_4_BYTES = 4;
+	private static final long STRUCT_ARG_SIZE_8_BYTES = 8;
+
+	private static final VaList EMPTY = new SharedUtils.EmptyVaList(emptyListAddress());
+
+	private final MemorySegment segment;
+	private final MemorySegment regSaveAreaOfVaList;
+	private final MemorySegment overflowArgAreaOfVaList;
+	private MemorySegment gpRegSaveArea;
+	private MemorySegment fpRegSaveArea;
+	private MemorySegment overflowAreaCursor;
+
+	private SysVS390xVaList(MemorySegment segment, MemorySegment gpRegSaveArea, MemorySegment fpRegSaveArea, MemorySegment overflowArgArea) {
+		this.segment = segment;
+		this.gpRegSaveArea = gpRegSaveArea;
+		this.fpRegSaveArea = fpRegSaveArea;
+		this.overflowAreaCursor = overflowArgArea;
+		this.regSaveAreaOfVaList = MemorySegment.ofAddress((MemoryAddress)VH_REG_SAVE_AREA.get(segment),
+													LAYOUT_REG_SAVE_AREA.byteSize(), segment.session());
+		this.overflowArgAreaOfVaList = MemorySegment.ofAddress((MemoryAddress)VH_OVERFLOW_ARG_AREA.get(segment),
+																Long.MAX_VALUE, segment.session());
+	}
+
+	private static MemoryAddress emptyListAddress() {
+		long vaListPtr = unsafe.allocateMemory(LAYOUT_VA_LIST.byteSize());
+		MemorySession session = MemorySession.openImplicit();
+		session.addCloseAction(() -> unsafe.freeMemory(vaListPtr));
+		MemorySegment vaListSegment = MemorySegment.ofAddress(MemoryAddress.ofLong(vaListPtr),
+																LAYOUT_VA_LIST.byteSize(), session);
+		VH_GPR_NO.set(vaListSegment, 0);
+		VH_FPR_NO.set(vaListSegment, 0);
+		VH_OVERFLOW_ARG_AREA.set(vaListSegment, MemoryAddress.NULL);
+		VH_REG_SAVE_AREA.set(vaListSegment, MemoryAddress.NULL);
+		return vaListSegment.address();
+	}
+
+	public static VaList empty() {
+		return EMPTY;
+	}
+
+	@Override
+	public int nextVarg(ValueLayout.OfInt layout) {
+		return Math.toIntExact((long)readArg(layout));
+	}
+
+	@Override
+	public long nextVarg(ValueLayout.OfLong layout) {
+		return (long)readArg(layout);
+	}
+
+	@Override
+	public double nextVarg(ValueLayout.OfDouble layout) {
+		return (double)readArg(layout);
+	}
+
+	@Override
+	public MemoryAddress nextVarg(ValueLayout.OfAddress layout) {
+		return (MemoryAddress)readArg(layout);
+	}
+
+	@Override
+	public MemorySegment nextVarg(GroupLayout layout, SegmentAllocator allocator) {
+		return (MemorySegment)readArg(layout, allocator, true);
+	}
+
+	@Override
+	public void skip(MemoryLayout... layouts) {
+		Objects.requireNonNull(layouts);
+		sessionImpl().checkValidState();
+		for (MemoryLayout layout : layouts) {
+			readArg(layout, THROWING_ALLOCATOR, false);
+		}
+	}
+
+	private Object readArg(MemoryLayout layout) {
+		return readArg(layout, THROWING_ALLOCATOR, true);
+	}
+
+	private Object readArg(MemoryLayout layout, SegmentAllocator allocator, boolean isRead) {
+		Objects.requireNonNull(layout);
+		Objects.requireNonNull(allocator);
+		TypeClass typeClass = TypeClass.classifyLayout(layout);
+		long nextGprNo = currentGprNo() + 1;
```

Comment:
Is is possible that the value returned by `currentGprNo()` here become stale when `nextGprNo` is used later in this function ?

Reply:
Nope, it won't happen as the access to VaList only works in a single thread as specified in the API document.